### DISCUSSION TITLE
Live preview when editing and page revisions/drafts

### DISCUSSION THREAD

**Discussion by kevin-invato** [OFFLINE-GmbH/oc-boxes-support#49] [Help]
Hey there,

We really like this plugin and see a lot of potential with it. We've been experimenting with the plugin and boxes possibilities and it seems perfect for our clients and could really improve the way our clients manage their website.

However, we do have some questions, and I'm sorry if these questions are already asked.

The thing we love to see is a live preview of the page. Like whenever a box form field is changed. Is this something we can build ourselves? If not, will you consider implementing this?

Another thing we love to see, and perhaps more important than the live preview is the possibility to have page revisioning. In case one of our clients makes a mistake, it would be helpful if they could restore the page. Same as the option to publish a page as draft (and set a publish date), like the tailor models.

I love to see your reaction and reply. If the features above would be possible, we are really considering using the plugin for all our clients in the future :-)

Best regards,

Kevin,
[Invato B.V.](https://octobercms.com/partners/profile/invato)

**Comment by tobias-kuendig** [comment]
Hey Kevin

Thank you very much for your feedback!

> The thing we love to see is a live preview of the page. Like whenever a box form field is changed. Is this something we can build ourselves? If not, will you consider implementing this?

I have added this to our internal issue tracker. We can use October's change tracking for fields and refresh the preview live.  We will have to see how this performs, though. There will be at least a few ms of delay, so the preview won't be _instant_.

> Another thing we love to see, and perhaps more important than the live preview is the possibility to have page revisioning. In case one of our clients makes a mistake, it would be helpful if they could restore the page. Same as the option to publish a page as draft (and set a publish date), like the tailor models.

This is absolutely a must have! We actually have noticed the need for this ourselfes. It is not ideal, that all changes are published right away. The reason this feature has not been implemented yet is that it is hard to get it right. A page can have a nested tree of boxes and each box can have multiple file attachments (and other relations) of various sizes. Having a proper revision system would mean that we have to copy each of these relations for every box with every page revision, which can be complicated to get right.

October has added great new replicating features for models and model relations in recent versions. So we might give this feature another try. I'll let you know as soon as a early version is available.

**Comment by tobias-kuendig** [comment]
> Another thing we love to see, and perhaps more important than the live preview is the possibility to have page revisioning.

If you want to, you can test the current progress of this feature by installing the 2.2.0 pre-release. Please install this version only in test projects as it may need some db resets with future updates until it is released.

```
composer require offline/oc-boxes-plugin:^2.2.0-alpha.2
```

**Comment by kevin-invato** [comment]
Hi Tobias, Thanks for your replies!

> We can use October's change tracking for fields and refresh the preview live. We will have to see how this performs, though. There will be at least a few ms of delay, so the preview won't be instant.

That sounds promising, and little delay is no problem at all. Although, I can imagine it could get a bit confusing if one focusses on the preview instead of the form field when typing, but we'll see and test it when the time comes.


> This is absolutely a must have! We actually have noticed the need for this ourselfes. It is not ideal, that all changes are published right away. 

Glad to hear we aren't the only ones feeling the need for that feature hehe. I can understand why this feature is a lot more complicated than it sounds and I'm happy to hear you are working on it.


> If you want to, you can test the current progress of this feature by installing the 2.2.0 pre-release. Please install this version only in test projects as it may need some db resets with future updates until it is released.

That's awesome (and quick)! It's really good to see that you guys are working hard on the plugin :-)

I will definitely test the pre-release asap!


Again, thanks for the reply. I will let you know my thoughts about the pre-release :-)
